librz/bin/format/pe/pe.c

```diff
@@ -403,12 +403,7 @@ ut64 PE_(rz_bin_pe_get_image_base)(struct PE_(rz_bin_pe_obj_t) * bin) {
 		return 0LL;
 	}
 	imageBase = bin->nt_headers->optional_header.ImageBase;
-	if (!imageBase) {
```

Comment:
IMHO rizin should done the best by default. We could check the system in `bin->nt_headers->optional_header.Subsystem` and if the value is a windows one we set the imagebase=0x10000, otherwise use just the binary value.

Reply:
If a file is bogus u can not rely on the subsystem flag either.

The test file is bogus. Windows does not load it.

Reply:
// but from here we can not access config API

Also this is the wrong location to handle this

Reply:
> If a file is bogus u can not rely on the subsystem flag either.

True but you can provide useful defaults for the good case. If you analyze a specially crafted binary you can still use `-B` when loading the binary.

> 
> The test file is bogus. Windows does not load it.

The test file is not bogus AFAIK because it's a binary from Corkami.

Reply:
// this should only happens with messed up binaries

if ( strcmp(filename, "virtrelocXP.exe") == 0) { ... }

Reply:
> 
> > The test file is bogus. Windows does not load it.
> 
> The test file is not bogus AFAIK because it's a binary from Corkami.

Look at it's disassembly.

 * It set ImageBase to 0x0 but "call [0x10208]" the offset was crafted.
 * It's first instruction "push 0x150" is wrong if we set ImageBase to 0x10000 (to my understanding should be 0x10150, 1st arg to printf?!)
 * windows raises middle finger (I tried it)
 * the file is "synthetic"
 
https://github.com/indrora/corkami/blob/master/src/PE/virtrelocXP.asm

Reply:
> > > The test file is bogus. Windows does not load it.
> > 
> > 
> > The test file is not bogus AFAIK because it's a binary from Corkami.
> 
> Look at it's disassembly.
> 
> * It set ImageBase to 0x0 but "call [0x10208]" the offset was crafted.
> * It's first instruction "push 0x150" is wrong if we set ImageBase to 0x10000 (to my understanding should be 0x10150, 1st arg to printf?!)

IDA disassembles that correctly. There is likely a relocation that relocates 0x150 to 0x10150.

> * windows raises middle finger (I tried it)

Yes, maybe there is some protection or new check in newer windows version that prevents that from loading. I bet if you spawn a Windows XP VM though it will run.

> * the file is "synthetic"
> 
> https://github.com/indrora/corkami/blob/master/src/PE/virtrelocXP.asm

The fact that it is synthetic does not really matter.


Anyway, I agree probably this is not the best place. I suggest the get_image_base in bin_pe plugin returns the value in the PE header, but somewhere else we set anyway the 0x10000 for windows binaries in case the user has not set anything else. I think we can rely on the value in RzBinObjectLoadOptions.baseaddr to check the baseaddr that is set.